docs: SIWE tutorial and Solidity polishes (#459)
oasisprotocol/sapphire-paratime: docs/authentication.md

    of any `from` overrides passed on the client side for simulating the query.
   
   :::note

   Calldata end-to-end encryption has nothing to do with authentication.
   Although the calls may be unauthenticated they can still be encrypted, and
   the other way around!

   :::

3. **Authenticated view calls** (via SIWE token)

    Developer authenticates the view call explicitly by deriving a message
    sender from the SIWE token. This token is provided as a separate parameter
    to the contract function. The derived address can then be used for
    authentication in place of `msg.sender`. Otherwise, such view call behaves
    the same way as the unauthenticated view calls above and built-in
    `msg.sender` is `address(0x0)`. This approach is most appropriate for 
    frontend dApps.

4. **Authenticated view calls** (via signed queries)

    [EIP-712] defines a format for signing view calls with the keypair of your
    Ethereum account. Sapphire will validate such signatures and automatically
    set the `msg.sender` parameter in your contract to the address of the
    signing account. This method is mostly appropriate for backend services,

>     since the frontend would require user interaction each time.

Reviewer: We don't want to nudge people towards 3?

Author: EIP-712 should work perfectly fine for backend services, so it really depends on the app. I deliberately put SIWE above the signed queries though.

        return _message;
    }

    function setSecretMessage(string calldata message) external onlyOwner {
        _message = message;
    }
}
```

### via SIWE token

SIWE stands for "Sign-In with Ethereum" and is formally defined in [EIP-4361].
The initial use case for SIWE was using your Ethereum account as a form of
authentication for off-chain services (as an alternative to the user names and
passwords). The MetaMask wallet quickly adopted the standard and it became a
de-facto login mechanism in the Web3 world. An informative pop-up for logging
into a SIWE-enabled website looks like this:

![MetaMask Log-In confirmation](images/siwe-login.png)

Once a user agrees with signing the SIWE login message above, the signature is
verified on the website backend or by a 3rd party [single sign-on] service. This
is done only once—during login. Then, a session token is generated and used for
the rest of the session.

In contrast to transparent EVM chains, **Sapphire simplifies dApp design,
improves trust and extends the usability of SIWE by supporting on-chain SIWE
message parsing and verification**. This unique feature removes the need for

> running a separate dApp backend just for SIWE authentication. Let's look at such

Reviewer: ```suggestion
running a separate dApp backend just for SIWE authentication. Let's take a look at this
```

  address private _owner;
  string private _message;

  modifier onlyOwner(bytes memory bearer) {
    if (msg.sender != author && authMsgSender(bearer) != author) {
        revert("not allowed");
    }
    _;
  }

  constructor(string memory domain) SiweAuth(domain) {
    _owner = msg.sender;
  }

  function getSecretMessage(bytes calldata bearer) external view onlyOwner(bearer) returns (string memory) {
    return _message;
  }

  function setSecretMessage(string calldata message) external onlyOwner(bytes("")) {
    _message = message;
  }
}
```

The following changes were made:

1. In the constructor, we need to define the domain name where the dApp frontend
   will be deployed at. This domain is included inside the SIWE log-in message
   and is verified by the user-facing wallet to make sure they are accessing the
   contract from a legit domain.
2. The `onlyOwner` modifier is extended with an optional `bytes memory bearer`
   parameter and is considered in the case of invalid `msg.sender` value. The
   same modifier is used for authenticating both SIWE queries and the
   transactions.
3. `getSecretMessage` was extended with the `bytes memory bearer` session token.

> On the client side, the TypeScript code running inside a browser needs to make

Author: I'll just omit the language name ;)

sure that the session token for making authenticated calls is valid. If not, the
browser requests a wallet to sign a log-in message and fetches a fresh session
token.

```typescript
import {SiweMessage} from 'siwe';

async function getSecretMessage(): Promise<Message> {
  const messageBox = await hre.ethers.getContractAt('MessageBox', '0x5FbDB2315678afecb367f032d93F642f64180aa3');

  if (token == '') { // Stored in browser session.
    const domain = await messageBox.domain();
    const siweMsg = new SiweMessage({
      domain,
      address: addr, // User's selected account address.
      uri: `http://${domain}`,
      version: "1",
      chainId: 0x5afe, // Sapphire Testnet
    }).toMessage();
    const sig = Signature.from((await window.ethereum.getSigner(addr)).signMessage(siweMsg));
    token = await messageBox.login(siweMsg, sig);
  }

>   return messageBox.getSecretMessage(bearer);

Author: I did some research and decided to go ahead with `token` everywhere since "Bearer" is associated with specially formatted token string according to RFC 6750.

}
```

:::info Example: Starter project

To see a running example of the TypeScript SIWE code including the Hardhat
tests, Node.js and the browser, check out the official Oasis [demo-starter]
project.

:::

:::tip Sapphire TypeScript wrapper?

While the [Sapphire TypeScript wrapper][sp-npm] offers a convenient end-to-end
encryption for the contract calls out of the box, it is not mandatory for SIWE.

:::

[demo-starter]: https://github.com/oasisprotocol/demo-starter/tree/matevz/sapphire-paratime-2.0
[EIP-4361]: https://eips.ethereum.org/EIPS/eip-4361
[single sign-on]: https://en.wikipedia.org/wiki/Single_sign-on
[sp-npm]: https://www.npmjs.com/package/@oasisprotocol/sapphire-paratime

### via signed queries

> The [EIP-712] proposal defines a method to show data to the user in a structured
> fashion so they can verify it and sign it. In the frontend apps signing a view
> call would require user interaction each time—sometimes even multiple times per
> page—which results in bad UX. Backend services on the other hand can have direct
> access to an Ethereum wallet without needing user interaction. This is possible
> because such services never access random websites, but execute only trusted
> code.

Reviewer: ```suggestion
The [EIP-712] proposal defines a method to show data to the user in a structured
fashion, so they can verify it and sign it. On the frontend, apps signing a view
call would require user interaction each time—sometimes even multiple times per
page—which is bad UX that frustrates users. Backend services on the other hand can have direct
access to an Ethereum wallet without needing user interaction. This is possible
because these services do not access unspecified websites, and only execute trusted
code.
```

> The Sapphire wrappers for [Go][sp-go] and [Python][sp-py] will make sure to
> **automatically sign any view calls** you make to a contract running on Sapphire
> using the proposed [EIP-712] method. Suppose we store a private key of an
> account used for signing the view calls inside a `PRIVATE_KEY` environment variable. The following snippets
> demonstrate how to trigger signed queries without any changes to the original
> `MessageBox` contract from [above](#authenticated-view-calls).

Reviewer: ```suggestion
The Sapphire wrappers for [Go][sp-go] and [Python][sp-py] will make sure to
**automatically sign any view calls** you make to a contract running on Sapphire
using the proposed [EIP-712] method. Suppose we want to store the private key of an
account used to sign the view calls inside a `PRIVATE_KEY` environment variable. The following snippets
demonstrate how to trigger signed queries without any changes to the original
`MessageBox` contract from [above](#authenticated-view-calls).
```

    addr := crypto.PubkeyToAddress(*sk.Public().(*ecdsa.PublicKey))

    wrappedClient, err := sapphire.WrapClient(c.Client, func(digest [32]byte) ([]byte, error) {
        return crypto.Sign(digest[:], sk)
    })
    if err != nil {
        return "", fmt.Errorf("unable to wrap backend: %v", err)
    }

    mb, err := messageBox.NewMessageBox(common.HexToAddress("0x5FbDB2315678afecb367f032d93F642f64180aa3"), wrappedClient)
    if err != nil {
        return "", fmt.Errorf("Unable to get instance of contract: %v", err)
    }
	
    msg, err := mb.GetSecretMessage(&bind.CallOpts{From: addr}) // Don't forget to pass callOpts!
    if err != nil {
        return "", fmt.Errorf("failed to retrieve message: %v", err)
    }
	
    return msg, nil
}
```

    :::info Example: Oasis starter in Go

    To see a running example of the Go code including the end-to-end encryption
    and signed queries check out the official [Oasis starter project for Go].

    :::
    </TabItem>
    <TabItem value="Python">
    Wrap the existing Web3 client by calling the
    [`wrap()`][wrapped-py] helper and provide the signing logic. Then,
    all subsequent view calls will be signed. For example:

```python

> from web3 import Web3

Reviewer 2: We might also add the compilation and deployment step in the function

```
import os
import json
from typing import Optional, Tuple
from web3 import Web3
from web3.middleware import construct_sign_and_send_raw_middleware
from eth_account.signers.local import LocalAccount
from eth_account import Account
from sapphirepy import sapphire
from solcx import compile_standard, install_solc

def deploy_and_get_message(
    solidity_file: str, 
    network_name: Optional[str] = "sapphire-localnet"
) -> Tuple[str, str]:
    # Initialize web3 and account
    w3 = Web3(Web3.HTTPProvider(sapphire.NETWORKS[network_name]))
    account: LocalAccount = Account.from_key(os.environ.get("PRIVATE_KEY"))
    w3.middleware_onion.add(construct_sign_and_send_raw_middleware(account))
    w3 = sapphire.wrap(w3)
    
    # Compile the Solidity file
    install_solc('0.8.0')  # Install specific solidity version
    with open(solidity_file, 'r') as file:
        solidity_code = file.read()
    
    compiled_contract = compile_standard({
        "language": "Solidity",
        "sources": {solidity_file: {"content": solidity_code}},
        "settings": {
            "outputSelection": {
                "*": {
                    "*": ["abi", "metadata", "evm.bytecode", "evm.sourceMap"]
                }
            }
        }
    }, solc_version="0.8.0")
    
    # Get contract binary and abi
    contract_name = solidity_file.split('/')[-1].split('.')[0]
    contract_data = compiled_contract["contracts"][solidity_file][contract_name]
    bytecode = contract_data["evm"]["bytecode"]["object"]
    abi = contract_data["abi"]
    
    # Deploy contract
    Contract = w3.eth.contract(abi=abi, bytecode=bytecode)
    nonce = w3.eth.get_transaction_count(account.address)
    
    # Submit deployment transaction
    transaction = Contract.constructor().build_transaction({
        "chainId": w3.eth.chain_id,
        "from": account.address,
        "nonce": nonce,
        "gasPrice": w3.eth.gas_price,
    })
    
    # Send transaction and wait for receipt
    tx_hash = w3.eth.send_transaction(transaction)
    tx_receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
    contract_address = tx_receipt.contractAddress
    
    # Create contract instance and get message
    contract = w3.eth.contract(address=contract_address, abi=abi)
    message = contract.functions.getMessage().call()
    
    return contract_address, message
```
